### ISSUE TITLE
Ability to break up switch statement into small functions?

### ISSUE THREAD

**Issue by nepx** [WebAssembly/binaryen#1592]
Hello, I'm currently writing an emulator and I'm playing with Binaryen to see if I can tweak my code to squeeze out as many instructions as possible (Currently, I'm compiling my code to WebAssembly using Emscripten). My question is if Binaryen has the ability to turn `case`s in `switch` statements into function tables. 

My emulator has two loops: one for instruction decoding and one for IR execution (the IR makes it easier to translate into native code/asm.js/WebAssembly/etc.). One of them goes like this:

```c
switch(opcode){
   case 0: // Decode an ADD instruction...
   case 8: // Decode an OR instruction...
   // etc......... for hundreds of more cases.......
}
```

and the other looks like:

```c
switch(opcode){
    case ADD: // Do some addition
    case OR: // Do some bitwise OR'ing
    // etc........... goes on and on and on....
}
```

Currently, my speed bottleneck is in the switch statement (it is much slower than native due to there being no computed `goto` in WebAssembly) and I would be interested to see if I could change them to use function pointers directly in Binaryen without having to mess with the source (because it's a macro-infested mess right now) or even experimenting with a mix of the two (using a function table inside of a switch statement or vice versa).

**Comment by kripken** [comment]
We don't currently have a pass to do that. It would be interesting to experiment with one, though!

The main problem though is stuff on the stack. Converting a big function into separate smaller functions means that any locals need to be stored in linear memory or passed around as extra parameters. That overhead might be worse than the switch overhead, as the switch might not be that bad - it's not a computed goto, but it is a jump table in most VMs.

If you don't have stuff on the stack, then this is more likely to help. However, even so, wasm function pointer calls may not be faster than a jump table. Actually the jump table may be better in general, I am told by @sunfishcode

**Comment by nepx** [comment]
I see. I was worried that browsers might deoptimize large switch statements (like [V8 in the past](https://github.com/petkaantonov/bluebird/wiki/Optimization-killers#4-switch-case)) but I guess they don't these days. 

A computed goto definiately helps performance as well. On native, the switch-case version ran at just 800,000 blocks per second while the computed goto ran at either 10 or 100 million blocks per second (I forgot which one), although a dynamic recompiler can definitely help there (one block contains more than one x86 instruction and several micro-ops, so performance isn't that dismal).

**Comment by kripken** [comment]
Yeah, in wasm br_table is always dense and simple, and I think VMs always implement it that way - unlike switches in JS where, as you said, browsers had all sorts of heuristics and deopts.

Would be good to see those measurements on wasm too - some slowdown compared to native is expected, but hopefully not that much.

**Comment by binji** [comment]
> I think VMs always implement it that way

For baseline compilers, I think you're right. But I know for example that v8's TurboFan based wasm compiler doesn't always use a jump table; there are some heuristics, but no deopts.

**Comment by nepx** [comment]
Well, I did some poking around and it turns out I forgot to turn on optimizations. 
I tested it out and it turns out that the browser version of the switch statement table is actually consistently *faster* than native (approx 10,500,000 blocks per second on Firefox vs. 10,000,000 on native, wow!). So I guess the browser is doing some neat optimizations under the hood and I shouldn't lose any sleep over this, although a computed goto would be really nice.

**Comment by torch2424** [comment]
Hello @nepx !

As in "forgot to turn on optimizations", you mean you forgot the `-O3` flag?

I'm also [building an emulator](https://github.com/torch2424/wasmBoy) and happened to stumble across this.

Also is your emulator open source anywhere? :)

I don't want to derail the original issue though, so feel free to reach out over email, Twitter, etc...

**Comment by nepx** [comment]
@Czy-Zak thanks for the interest, but the project went nowhere. Seven years ago when I wrote that comment, I had significantly less experience writing CPU emulators and code generators, and in the end it was never able to run Hello World. 

I'm thinking about userspace emulation to my [full system x86 emulator](https://github.com/nepx/halfix), but that's for the future. It would be a good way to run test cases. 

Bringing the conversation back to the issue, one thing that's available now that wasn't back in 2018 are [WebAssembly tail calls](https://github.com/WebAssembly/tail-call/blob/main/proposals/tail-call/Overview.md), making this feature obsolete. In any case, any sufficiently complicated CPU will cause the execution loop switch statement to become so large that you might as well write small functions to begin with, so I think it would be best to close this issue for now.
